File: Java/백준/Gold/2239. 스도쿠/스도쿠.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {
    static int[][] sudoku = new int[9][9];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        for (int i = 0; i < 9; i++) {
            String line = br.readLine();
            for (int j = 0; j < 9; j++) {
                sudoku[i][j] = line.charAt(j) - '0';
            }
        }

        solve(0, 0);
    }

    static void solve(int row, int col) {
        if (col == 9) {
            solve(row + 1, 0);
            return;
        }

        if (row == 9) {
            printSudoku();
            System.exit(0);
        }

        if (sudoku[row][col] == 0) {
            for (int num = 1; num <= 9; num++) {
                if (isValid(row, col, num)) {
                    sudoku[row][col] = num;
                    solve(row, col + 1);
                    sudoku[row][col] = 0;
                }
            }
        } else {
            solve(row, col + 1);
        }
    }

    static boolean isValid(int row, int col, int num) {
        for (int i = 0; i < 9; i++) {
            if (sudoku[row][i] == num || sudoku[i][col] == num) {
                return false;
            }
        }

        int startRow = (row / 3) * 3;
        int startCol = (col / 3) * 3;

        for (int i = startRow; i < startRow + 3; i++) {
            for (int j = startCol; j < startCol + 3; j++) {
                if (sudoku[i][j] == num) {
                    return false;
                }
            }
        }

        return true;
    }

    static void printSudoku() {
        StringBuilder sb = new StringBuilder();
        for (int[] row : sudoku) {
            for (int val : row) {
                sb.append(val);
            }
            sb.append("\n");
        }
        System.out.print(sb);
    }
}

File: C11/백준/Gold/2239. 스도쿠/스도쿠.c
#include <stdio.h>
#include <stdlib.h>

int sudoku[9][9];

int is_valid(int row, int col, int num) {
    for (int i = 0; i < 9; i++) {
        if (sudoku[row][i] == num || sudoku[i][col] == num)
            return 0;
    }

    int start_row = (row / 3) * 3;
    int start_col = (col / 3) * 3;

    for (int i = start_row; i < start_row + 3; i++) {
        for (int j = start_col; j < start_col + 3; j++) {
            if (sudoku[i][j] == num)
                return 0;
        }
    }

    return 1;
}

void print_sudoku() {
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            printf("%d", sudoku[i][j]);
        }
        printf("\n");
    }
    exit(0);
}

void solve(int row, int col) {
    if (col == 9) {
        solve(row + 1, 0);
        return;
    }

    if (row == 9) {
        print_sudoku();
    }

    if (sudoku[row][col] == 0) {
        for (int num = 1; num <= 9; num++) {
            if (is_valid(row, col, num)) {
                sudoku[row][col] = num;
                solve(row, col + 1);
                sudoku[row][col] = 0;
            }
        }
    } else {
        solve(row, col + 1);
    }
}

int main() {
    for (int i = 0; i < 9; i++) {
        char line[10];
        scanf("%s", line);
        for (int j = 0; j < 9; j++) {
            sudoku[i][j] = line[j] - '0';
        }
    }

    solve(0, 0);
    return 0;
}